pkg/embeddedcluster/upgrade.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/replicatedhq/kots/pkg/k8sutil"
+	kotsv1beta1 "github.com/replicatedhq/kotskinds/apis/kots/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+const (
+	clusterUpgradeTimeout = 5 * time.Minute
+)
+
+// startClusterUpgrade will create a new installation with the provided config.
+func startClusterUpgrade(ctx context.Context, newcfg embeddedclusterv1beta1.ConfigSpec, artifacts *embeddedclusterv1beta1.ArtifactsLocation, license kotsv1beta1.License) error {
+	// TODO(upgrade): put a lock here to prevent multiple upgrades at the same time
+
+	kbClient, err := k8sutil.GetKubeClient(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get kubeclient: %w", err)
+	}
+	current, err := GetCurrentInstallation(ctx, kbClient)
+	if err != nil {
+		return fmt.Errorf("failed to get current installation: %w", err)
+	}
+	newins := embeddedclusterv1beta1.Installation{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: embeddedclusterv1beta1.GroupVersion.String(),
+			Kind:       "Installation",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: time.Now().Format("20060102150405"),
+			Labels: map[string]string{
+				"replicated.com/disaster-recovery": "ec-install",
+			},
+		},
+		Spec: embeddedclusterv1beta1.InstallationSpec{
+			ClusterID:                 current.Spec.ClusterID,
+			MetricsBaseURL:            current.Spec.MetricsBaseURL,
+			HighAvailability:          current.Spec.HighAvailability,
+			AirGap:                    current.Spec.AirGap,
+			Artifacts:                 artifacts,
+			Config:                    &newcfg,
+			EndUserK0sConfigOverrides: current.Spec.EndUserK0sConfigOverrides,
+			BinaryName:                current.Spec.BinaryName,
+			LicenseInfo:               &embeddedclusterv1beta1.LicenseInfo{IsDisasterRecoverySupported: license.Spec.IsDisasterRecoverySupported},
+		},
+	}
+
+	installationData, err := k8syaml.Marshal(newins)
+	if err != nil {
+		return fmt.Errorf("failed to marshal installation: %w", err)
+	}
+
+	log.Printf("Upgrading cluster to version %s...", newcfg.Version)
+
+	err = runClusterUpgrade(ctx, installationData)
+	if err != nil {
+		return fmt.Errorf("run cluster upgrade: %w", err)
+	}
+
+	log.Printf("Cluster upgrade to version %s completed successfully", newcfg.Version)
+
+	return nil
+}
+
+const (
+	// TODO(upgrade)
+	localArtifactMirrorImage = "ttl.sh/ethan/embedded-cluster-local-artifact-mirror:24h"
+)
+
+// runClusterUpgrade will download the new embedded cluster operator binary and run the upgrade
+// command with the provided installation data. This is needed to get the latest
+// embeddedclusterv1beta1 API version. The upgrade command will first upgrade the embedded cluster
+// operator, wait for the CRD to be up-to-date, and then apply the installation object.
+func runClusterUpgrade(ctx context.Context, installationData []byte) error {
+	log.Println("Downloading upgrade binary...")
+
+	bin, err := downloadUpgradeBinary(ctx)
```

Comment:
This needs to account for airgap and pull the operator binary from the registry.

Reply:
done